### PR TITLE
[__implements__] Remove everything related to the rejected PEP245

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Everything related to the ``__implements__`` construct was removed. It was based on PEP245
that was proposed in 2001 and rejected in 2006. It includes everything in ``pylint.interfaces``

``Checker`` should only inherit ``BaseChecker`` or any of the other checker types from ``pylint.checkers``.
``Reporter`` should only inherit ``BaseReporter``.

The capability from pyreverse to take it into account when generating diagrams was also removed.

### CODE REVIEW

File: pylint/pyreverse/inspector.py [hunk 1]
```diff
@@ -194,24 +171,7 @@ def visit_classdef(self, node: nodes.ClassDef) -> None:
...
-                        "implemented interfaces in pylint 3.0. The implementation "
-                        "relies on the '__implements__'  attribute proposed in PEP 245"
-                        ", which was rejected in 2006.",
-                        DeprecationWarning,
-                    )
-            else:
-                node.implements = []
-        except astroid.InferenceError:
-            node.implements = []
+        node.implements = []
```
Comment (DanielNoord) [inline pylint/pyreverse/inspector.py:174]: Can't we just remove this attribute as well?

Comment (Pierre-Sassoulas) [reply pylint/pyreverse/inspector.py:174]: I was wondering also if you were planning on removing all ``IChecker`` classes ?

Comment (DanielNoord) [reply pylint/pyreverse/inspector.py:174]: No, they just don't need to be `Ichecker` anymore

Comment (DanielNoord) [reply pylint/pyreverse/inspector.py:174]: Oh wait, they do need to be. So no, they should remain as is.

Comment (DanielNoord) [reply pylint/pyreverse/inspector.py:174]: Wait.... hahah Which `IChecker` classes are you referring to?

**Comment by DudeNr33** [comment]
For `pyreverse` there is some more code that can be cleaned up together with this. 
I can push to this branch tomorrow evening, or if you want to merge beforehand create a follow-up PR.

**Comment by Pierre-Sassoulas** [comment]
I've done the cleanup, I'm going to push :)

File: pylint/checkers/base_checker.py [hunk 2]
```diff
@@ -193,21 +184,10 @@ def check_consistency(self) -> None:
...
-                warnings.warn(  # pragma: no cover
-                    "Checkers should subclass BaseTokenChecker or BaseRawFileChecker "
-                    "instead of using the __implements__ mechanism. Use of __implements__ "
-                    "will no longer be supported in pylint 3.0",
-                    DeprecationWarning,
-                )
-                default_scope = WarningScope.LINE  # pragma: no cover
-            else:
-                default_scope = WarningScope.NODE
+        if isinstance(self, (BaseTokenChecker, BaseRawFileChecker)):
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/base_checker.py:187]:
```suggestion
        if isinstance(self, (BaseTokenChecker, BaseRawFileChecker, IRawChecker, ITokenChecker):
```
Should we do this here @DanielNoord ?

Comment (DanielNoord) [reply pylint/checkers/base_checker.py:187]: No I was mistaken. Those should be removed. Everything that gets tested for Deprecation should be removed indeed!

File: tests/test_deprecation.py [hunk 3]
```diff
@@ -44,45 +33,6 @@ def reduce_map_data(self, linter: PyLinter, data: list[Any]) -> None:
...
-    with pytest.warns(DeprecationWarning):
-        IAstroidChecker()
-    with pytest.warns(DeprecationWarning):
-        IReporter()
-    with pytest.warns(DeprecationWarning):
-        IRawChecker()
-    with pytest.warns(DeprecationWarning):
-        IChecker()
-    with pytest.warns(DeprecationWarning):
-        ITokenChecker()
```
[the remark this answers is not in the document]
Comment (DanielNoord) [reply tests/test_deprecation.py:83]: Yeah I was looking at the wrong code. This should definitely be removed! Sorry for the confusion!

**Review by DudeNr33 — APPROVED** [review]
There are a few more places in `pyreverse` that can be cleaned up now, but the important stuff is included and tests adapted, thanks for that!

From the first quick look the following things can be removed as well:

* `DiagramWriter.write_classes` --> everything related to implementation links
* `printer.py`: `NodeType.INTERFACE` and `EdgeType.IMPLEMENTS`, and for the concrete `Printer` classes the places where those are used

As said I can take a closer look tomorrow and create a follow up PR. Especially the `NodeType.INTERFACE` is something that as far as I know did not even work as intended.

**Review by DanielNoord — APPROVED** [review]

**Review by DanielNoord — APPROVED** [review]

**Comment by nnnnnnsssss** [comment]
Recheck

**Pull request merged**